sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,240 @@
+
+        /// <summary>
+        /// Last known length of underlying blob or file.
+        /// </summary>
+        private long _length;
+
+        /// <summary>
+        /// The number of bytes in the last download call.
+        /// </summary>
+        private int _lastDownloadBytes;
+
+        /// <summary>
+        /// The number of bytes to download per call.
+        /// </summary>
+        private readonly int _bufferSize;
+
+        /// <summary>
+        /// The stream we will download to.
+        /// </summary>
+        private Stream _stream;
+
+        /// <summary>
+        /// Request conditions to send on the download requests.
+        /// </summary>
+        private readonly object _requestConditions;
+
+        /// <summary>
+        /// Async DownloadTo() function.
+        /// </summary>
+        private readonly Func<HttpRange, object, bool, CancellationToken, Task<Response<T>>> _downloadToAsyncFunc;
+
+        /// <summary>
+        /// Sync DownloadTo() function.
+        /// </summary>
+        private readonly Func<HttpRange, object, bool, CancellationToken, Response<T>> _downloadToFunc;
+
+        public LazyLoadingReadOnlyStream(
+            Func<HttpRange, object, bool, CancellationToken, Task<Response<T>>> downloadToAsyncFunc,
+            Func<HttpRange, object, bool, CancellationToken, Response<T>> downloadToFunc,
+            long position = 0,
+            int bufferSize = Constants.DefaultDownloadCopyBufferSize,
+            object requestConditions = default)
+        {
+            _downloadToAsyncFunc = downloadToAsyncFunc;
+            _downloadToFunc = downloadToFunc;
+            _position = position;
+            _bufferSize = bufferSize;
+            _stream = new MemoryStream(bufferSize);
+            _requestConditions = requestConditions;
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadInternal(
+                async: false,
+                buffer,
+                offset,
+                count,
+                default)
+            .EnsureCompleted();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => await ReadInternal(
+                async: true,
+                buffer,
+                offset,
+                count,
+                cancellationToken)
+                .ConfigureAwait(false);
+
+        public async Task<int> ReadInternal(bool async, byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateReadParameters(buffer, offset, count);
+
+            if (_stream.Position == 0)
+            {
+                await Download(async, cancellationToken).ConfigureAwait(false);
+                if (_lastDownloadBytes == 0)
+                {
+                    return 0;
+                }
+            }
+
+            int totalCopiedBytes = 0;
+            do
```

Comment:
I would have expected this to look more like PartitionedDownloader.  I think this would perform better with rolling buffers and fetching the next eagerly as soon as you read past the end of any.

Reply:
Maybe in the next version.